### PR TITLE
HW #3 (Tests)

### PR DESCRIPTION
**Pull request by Svitlana-Marchenko**
Key changes:
- add unit test for controllers and services
- add integration test for service

### CODE REVIEW

File: src/test/controller/user.controller.test.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    });
+
+    describe('test subscribeEmail method', () => {
+
+        it('should create a new user if user with given email does not exist', async () => {
+            req.body = { email: 'test@example.com' };
+            userServiceStub.subscribeEmail.resolves({ id: '1234', email: 'test@example.com' });
+
+            await UserController.subscribeEmail(req as Request, res as Response);
+
```
Comment (PavloVlasiuk) [inline src/test/controller/user.controller.test.ts:38]: you can also check if userService method subscribeEmail was called with correct params

File: src/test/service/user.service.test.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    afterEach(() => {
+        sinon.restore();
+    });
+
+    describe('test subscribeEmail method', () => {
+        it('should create a new user if user with given email does not exist', async () => {
+            userRepositoryStub.findOne.resolves(null);
+            userRepositoryStub.create.returns({ id: '1234', email: 'test@example.com' });
+            userRepositoryStub.save.resolves({ id: '1234', email: 'test@example.com' });
```
Comment (PavloVlasiuk) [inline src/test/service/user.service.test.ts:22]: create constant for testing email

**Review by PavloVlasiuk — CHANGES_REQUESTED** [review]

File: src/controllers/user.controller.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+                case error instanceof UserAlreadyExistsError:
+                    logger.error(`Email ${email} is already in the database`)
+                    return res.status(409).json({
+                        message: `Email ${email} is already in the database`,
+                    })
+                case error instanceof Error:
+                    logger.error(
+                        `Error adding email ${email} to the database: ${error.message}`
+                    )
+                    return res.status(500).json({error: error.message})
```
Comment (trndo) [inline src/controllers/user.controller.ts:33]: I suggest to handle this type of errors inside the service level. You should handle validation like errors in controllers.  Cause in the future your user service can be 3rd party service or cache and this part of code can be misleading

File: src/test/controller/user.controller.test.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            expect(res.status).toHaveBeenCalledWith(200);
```
[the remark this answers is not in the document]
Comment (trndo) [reply src/test/controller/user.controller.test.ts:39]: not always, it is requirement for REST approach

**Review by trndo — APPROVED** [review]
Good work 👍

**Comment by Svitlana-Marchenko** [comment]
Thank you for the review and helpful comments

File: src/controllers/user.controller.ts (hunk 3 above)
Comment (Svitlana-Marchenko) [reply src/controllers/user.controller.ts:33]: do you mean smth like
```
try {
    SERVICE LAYER WAS CALLED
    return res.status(200).json({
        message: `New email was added to the database`,
    })
} catch (error) {
    if (error instanceof UserAlreadyExistsError) {
        logger.error(`Email ${email} is already in the database`)
        return res.status(409).json({
            message: `Email ${email} is already in the database`,
        })
    } else {
        logger.error(
            `Error adding email ${email} to the database: ${error.message}`
        )
        return res.status(500).json({error: error.message})
    }
}
```
?

**Pull request merged**